File: leetcode/add_two_numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
    public:
        ListNode *root = NULL;

        ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
            int res;
            int overflow = 0;
            ListNode *cur = root;
            while (l1 || l2 || overflow) {
                // determine sum of digits at this position
                res = 0;
                if (l1 && l2) {
                    // Both numbers have a digit here
                    res = l1->val + l2->val;
                } else if (l1) {
                    // Only first number has digit here
                    res = l1->val;
                } else if (l2) {
                    // Only second number has digit here
                    res = l2->val;
                }
                res += overflow;

                if (res > 9) {
                    // Overflow occurred
                    overflow = 1;
                    res -= 10;
                } else {
                    // No overflow occurred
                    overflow = 0;
                }

                // Add this position to resultant list
                if (!cur) {
                    cur = new ListNode(res);
                    root = cur;
                } else {
                    cur->next = new ListNode(res);
                    cur = cur->next;
                }

                // Iterate
                if (l1) {
                    l1 = l1->next;
                }
                if (l2) {
                    l2 = l2->next;
                }
            }

            return root;
        }
};

File: leetcode/neetcode_roadmap/add_two_numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def __init__(self):
        self.head = None
        self.tail = None
        self.prev = None


    def add(self, new):
        if self.head is None:
            self.head = new
            self.tail = new
        else:
            self.tail.next = new
            self.tail = new


    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        carry = 0
        while l1 or l2:
            if l1 and l2:
                res = l1.val + l2.val + carry
                if res > 9:
                    res -= 10
                    carry = 1
                else:
                    carry = 0

                l1 = l1.next
                l2 = l2.next
            elif l1:
                res = l1.val + carry
                if res > 9:
                    res -= 10
                    carry = 1
                else:
                    carry = 0
                l1 = l1.next
            else:
                res = l2.val + carry
                if res > 9:
                    res -= 10
                    carry = 1
                else:
                    carry = 0
                l2 = l2.next

            new = ListNode(res, None)
            self.add(new)

        if carry == 1:
            new = ListNode(1, None)
            self.add(new)

        return self.head
